articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,268 @@
+---
+layout: default
+title: Cross-compiling ROS 2 packages
+abstract:
+  This article is a design proposal for extending the build tools for ROS 2 to
+  simplify cross-compiling ROS 2 packages for platforms currently not supported
+  by ROS 2.
+author: >
+  [Thomas Moulard](https://github.com/thomas-moulard),
+  [Juan Rodriguez Hortala](https://github.com/juanrh)
+published: false
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
+and the [steve/ros2_raspbian_tools repository](https://github.com/esteve/ros2_raspbian_tools)
+contain instructions for cross-compiling ROS 2 to unsupported platforms like
+ARM64 and ARM-HF, using some configuration files for CMake and Docker.
+Other projects like [steve/ros2_raspbian_tools](https://github.com/esteve/ros2_raspbian_tools),
+[esteve/ros2_java](https://github.com/esteve/ros2_java) provide support
+for additional platforms like Android and iOS.
+Ideally, we should be able to cross-compile a ROS package for ROS 2 by
+launching a single command.
+That command should also be implemented using an extensible
+design that allows supporting new platforms with low effort.
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation
+tutorial, based on building a sysroot for the target platform using QEMU and
+Docker, and then using [`cmake-toolchains`](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html)
+with a toolchain file pointing `CMAKE_SYSROOT` to the sysroot location, to
+compile with C and C++ cross-compilers.
+
+We propose adding a set of commands to wrap the instructions on the
+cross-compilation tutorial.
+Those commands would use a new workspace
+directory for the target platform, determined by a convention based on
+the target ECU, operating system, rmw implementation, and ROS distribution.
+Let us call that directory the _cc-root_.
+Under that convention, the ECU
+implies the architecture, but generic ECUs like `generic_armhf` would also be
+available. For example `generic_armhf-ubuntu_bionic-fastrtps-crystal` is an
+example _platform identifier_.
+Specifically, we would add the following commands:
+
+- A new `cc-setup-sysroot` command would use Docker to generate the sysroot
+  for the target platform, and store it in a `sysroot` subdirectory of the
+  cc-root.
+  This command would take arguments to specify the target platform,
+  for example:
+
+```bash
+cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic \
+                 --rmw fastrtps --distro crystal
```

Comment:
```suggestion
                 --rmw fastrtps --rosdistro crystal
```

--rosdistro, abbreviated `-r` would be more consistent with other tools in the ROS community and less likely to be misinterpreted as a Linux ("distro") distribution.

Reply:
An rmw can consist of more than the underlying DDS mechanism. For example there are static and dynamic typesupport variants of rmw_fastrtps. On some level only the underlying middleware is significant for setting up the sysroot as it defines what's available for rmw implementation packages to be built on top of it, but I do want to make sure that we're not equating **one** middleware provider with **one** rmw implementation.